File: jwt/ts/src/util/parser.ts
import { createHmac } from "node:crypto";

const getBase64String = (obj: Record<string, any>): string => {
	return Buffer.from(JSON.stringify(obj)).toString("base64url");
};

type GenerateHmacSignature = {
	secret: string;
	header: string;
	payload: string;
};

const generateHmacSignature = ({
	secret,
	header,
	payload,
}: GenerateHmacSignature): string => {
	const hmac = createHmac("sha256", secret);
	const signature = hmac.update(`${header}.${payload}`).digest("base64url");

	return signature;
};

export { generateHmacSignature, getBase64String };

File: jwt/python/util/parser.py
import base64
import hashlib
import hmac
from json import dumps
from typing import Any, Dict


def get_base64_str(obj: Dict[str, Any]) -> str:
    obj_bytes = dumps(obj).encode()
    encoded_bytes = base64.urlsafe_b64encode(obj_bytes)
    encoded_str = encoded_bytes.rstrip(b"=").decode()

    return encoded_str


def generate_hmac_signature(secret: str, header: str, payload: str) -> str:
    digest = hmac.new(
        key=secret.encode(),
        msg=f"{header}.{payload}".encode(),
        digestmod=hashlib.sha256,
    ).digest()
    signature = base64.urlsafe_b64encode(digest).rstrip(b"=").decode()

    return signature
